src/core/lib/surface/channel_init.h

```diff
@@ -19,13 +19,38 @@
 #ifndef GRPC_CORE_LIB_SURFACE_CHANNEL_INIT_H
 #define GRPC_CORE_LIB_SURFACE_CHANNEL_INIT_H
 
+#include <limits.h>
+
 #include <grpc/support/port_platform.h>
 
 #include "src/core/lib/channel/channel_stack_builder.h"
 #include "src/core/lib/surface/channel_stack_type.h"
 #include "src/core/lib/transport/transport.h"
 
-#define GRPC_CHANNEL_INIT_BUILTIN_PRIORITY 10000
+// Priority for channel registration functions to be used in
+// grpc_channel_init_register_stage().  The priority dictates the
+// order in which the registration functions run.
+//
+// When used to register a filter, the filter can either be appended or
+// prepended, thus dictating whether the filter goes at the top or bottom of
+// the stack. Higher priority functions can get closer to the top or bottom
+// of the stack than lower priority functions.
+enum {
+  // Default level. Most of filters should use this level if their location in
+  // the stack does not matter.
+  GRPC_CHANNEL_INIT_PRIORITY_LOW = 0,
+  // For filters that should be added after the group of filters with default
+  // priority, such as auth filters.
+  GRPC_CHANNEL_INIT_PRIORITY_MED = 10000,
+  // For filters that need to be close to top or bottom, such as protocol-level
+  // filters (client_authority, http-client, http-server) and stats (census).
+  GRPC_CHANNEL_INIT_PRIORITY_HIGH = 20000,
+  // For filters that need to be very close to the wire or surface.
+  GRPC_CHANNEL_INIT_PRIORITY_VERY_HIGH = 30000,
+  // For things that have to happen last, such as connected channel filter or
+  // surface server filter. Consider as reserved for gRPC internals.
+  GRPC_CHANNEL_INIT_PRIORITY_MAX = INT_MAX
```

Comment:
I feel like MAX should be a special case though. connected, client_channel, server. All of these things NEED to be bottom, and the stack won't work if they are not.

Though, IIRC they all have their own mechanisms of ensuring they are bottome, so maybe we are ok

Reply:
MAX is not special.  For example, we use MAX to both prepend the server filter and append the connected_channel filter.  There are two dimensions here, each at a slightly different layer of abstraction, and there's no easy way to check both of them in the same place.

As you said, they all have their own mechanisms for ensuring that they're in the right place, so I think we're okay without this.

Reply:
I am not so worried about the two filters on MAX because first this has been always the case and second, this is an internal mechanism and rather than some public API. People can potentially get it wrong and that is okay as things will break early.